### PR TITLE
Initial structure

### PR DESCRIPTION
**Pull request by cedrickmandocdoc**
PR for #6 
Hi this is just file structure proposal, I also removed dependencies that are not needed for now.
start:dev - start dev environment
start:prod - start prod environment

The main problem is the config. I think the config should be globally available and isolated to nest implementation the flow that I am thinking is that after npm start the config should be available and ready to be used by nest framework.  Something like this https://www.npmjs.com/package/config.
We really don't want the config the be bound on nest because of its lifecycle, the structure of the app will loose. since probably we will need it even before the nest app will start. For now I just require it

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -6,54 +6,28 @@
...
-    "start:dev": "nodemon",
-    "prestart:prod": "rm -rf dist && tsc",
-    "start:prod": "node dist/main.js",
-    "start:hmr": "node dist/server",
-    "test": "jest",
-    "test:cov": "jest --coverage",
-    "test:e2e": "jest --config ./test/jest-e2e.json",
-    "webpack": "webpack --config webpack.config.js"
+    "start:dev": "NODE_ENV=development nodemon",
+    "start:prod": "NODE_ENV=production npm run build && node ./build/main.js"
```
Comment (jmaicaaan) [inline package.json]: no script for `npm run build`

File: package.json [hunk 2]
```diff
@@ -6,54 +6,28 @@
...
-    "ts-jest": "^21.2.4",
-    "ts-loader": "^4.1.0",
-    "ts-node": "^4.1.0",
-    "tsconfig-paths": "^3.1.1",
-    "tslint": "5.3.2",
-    "webpack": "^4.2.0",
-    "webpack-cli": "^2.0.13",
-    "webpack-node-externals": "^1.6.0"
-  },
-  "jest": {
```
Comment (jmaicaaan) [inline package.json]: Why did you remove jest for unit testing?

File: src/app/database/database.provider.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  provide: 'DbConnectionToken',
+  useFactory: async () =>
+    await createConnection({
+      type: config.database.type,
+      host: config.database.host,
+      port: config.database.port,
+      username: config.database.username,
+      password: config.database.password,
+      database: config.database.database,
+      synchronize: true,
```
Comment (jmaicaaan) [inline src/app/database/database.provider.ts]: Add a `config.database.synchronize` to avoid undesirable side-effect during each environment

File: package.json (hunk 2 above)
Comment (jmaicaaan) [reply package.json]: I think we can start a fresh unit test even with this config 👍

File: src/app/user/user.controller.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import { Controller, Get, Inject } from '@nestjs/common';
+
+import { UserRepository } from './user.repository';
+
+@Controller('users')
+export class UserController {
+
+  constructor(
+    @Inject('UserRepositoryToken')
+    public readonly userRepository: UserRepository
```
Comment (jmaicaaan) [inline src/app/user/user.controller.ts]: Let's make this private class scoped dependency. We aren't really instantiating a controller but as a practice, we want the dependency to be class scoped only right? 😄

File: src/main.ts [hunk 5]
```diff
@@ -1,8 +1,18 @@
...
 async function bootstrap() {
-  const app = await NestFactory.create(AppModule);
-  await app.listen(3000);
+  const config = require(`../config/config.${process.env.NODE_ENV}.json`);
+  const nestApplication = await NestFactory.create(AppModule);
+
+  // setup the application
+  setupNestApplication(nestApplication);
+
+  await nestApplication.init();
```
Comment (jmaicaaan) [inline src/main.ts]: `nestApplication.init` can be not called already I think. It is because nestApplication.listen already initiating the nest application

File: src/main.ts [hunk 6]
```diff
@@ -1,8 +1,18 @@
 import { NestFactory } from '@nestjs/core';
-import { AppModule } from './app.module';
+import { AppModule } from './app/app.module';
+
+import { setupNestApplication } from './setup';
 
 async function bootstrap() {
-  const app = await NestFactory.create(AppModule);
-  await app.listen(3000);
+  const config = require(`../config/config.${process.env.NODE_ENV}.json`);
```
Comment (jmaicaaan) [inline src/main.ts]: In the next sprint iteration, let's create a config service to load the config. Requiring it without proper handling can cause it to crash

File: src/app/database/database.service.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
+import { Inject, Injectable } from '@nestjs/common';
+
+import { Connection } from 'typeorm';
+
+@Injectable()
+export class DatabaseService {
+  public constructor(
```
Comment (jmaicaaan) [inline src/app/database/database.service.ts]: Remove public modifier in the constructor

File: src/app/database/database.provider.ts [hunk 8]
```diff
@@ -0,0 +1,18 @@
+import { createConnection } from 'typeorm';
+
+const config = require(`../../../config/config.${process.env.NODE_ENV}.json`);
```
Comment (jmaicaaan) [inline src/app/database/database.provider.ts]: See my comment in the `main.ts` regarding config. Let me know your thoughts

File: src/app/database/database.provider.ts [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+
+export const DatabaseProvider = {
+  provide: 'DbConnectionToken',
+  useFactory: async () =>
+    await createConnection({
```
Comment (jmaicaaan) [inline src/app/database/database.provider.ts]: Can we just pass the `config.database` object?

File: package.json (hunk 1 above)
Comment (cedrickmandocdoc) [reply package.json]: ok I will add a build script

File: package.json (hunk 2 above)
Comment (cedrickmandocdoc) [reply package.json]: can we create a ticket adding unit test? I think we are starting from scratch, so I removed things not needed for the initial develop. then we create a ticket for that and we focus on that.

File: src/main.ts (hunk 6 above)
Comment (cedrickmandocdoc) [reply src/main.ts]: I think configService is not bad. but actually we need the config just even before the app load or even before the maint.ts will run. and the config will load right after the npm script. I think we can use something like this https://www.npmjs.com/package/config we will probably have a wrapper for http init and process exiting I think.

File: package.json [hunk 10]
```diff
@@ -6,42 +6,34 @@
...
-    "start:dev": "nodemon",
-    "prestart:prod": "rm -rf dist && tsc",
-    "start:prod": "node dist/main.js",
-    "start:hmr": "node dist/server",
-    "test": "jest",
-    "test:cov": "jest --coverage",
-    "test:e2e": "jest --config ./test/jest-e2e.json",
-    "webpack": "webpack --config webpack.config.js"
+    "build": "rm -rf build && tsc",
+    "test": "NODE_ENV=test jest --env=node",
```
Comment (jmaicaaan) [inline package.json]: Why do we need the `--env=node`?

File: tslint.json [hunk 11]
```diff
@@ -49,7 +46,10 @@
     ],
     "one-variable-per-declaration": [
       false
-    ]
+    ],
+    "trailing-comma": false,
+    "no-console": false,
+    "no-var-requires": false
```
Comment (jmaicaaan) [inline tslint.json]: Let's make the `no-var-requires` to true. We don't need it in the TS world

File: package.json (hunk 10 above)
Comment (cedrickmandocdoc) [reply package.json]: Hmm I am not sure I think ci is prompting a security error on accessing window object, not sure on this one
